### ISSUE TITLE
Bosh prepare pulls its singularity image in the CWD instead of the final destination

### ISSUE THREAD

**Issue by prioux** [boutiques/boutiques#515]
This is a followup to #405 and #404 , which are both closed.

The `bosh prepare` command, when pulling a singularity image, always does the actual pull in the current working directory instead of the target directory of the value of the argument of `--imagepath`.

E.g.

```
bosh exec prepare --imagepath /a/b/c/image.simg descriptor.json
```

will:

1) create lock directory `/a/b/c/image.simg-lock`  (that is OK)
2) start pulling the singularity image with `singularity pull --name image.simg.tmp shub://blah` (that is NOT OK)
3) attempt to rename the pulled image using the full pathnames:
  `os.rename("/a/b/c/image.simg.tmp", "/a/b/c/image.simg"` (that is NOT OK)

Step 2 is wrong because the name given is just the basename
Step 3 crashes because the image pulled at step 2 is in the wrong location, and so the paths are incorrect

Also, step 3 cannot simply assume that the image pulled locally can be moved to the target location with `os.rename(basename,full_path)`, because os.rename() will not work across filesystems. Besides, the invoker might have a very good reason to want the image to be in a specific external location, for instance the current working directory might not be writable.

**Comment by prioux** [comment]
I have pipeline in CBRAIN that are crashing because of this, still. I had forgotten all about the problem, yet when a CBRAIN user tried to launch a task, I had to re-investigate it all and eventually found my way back here.

**Comment by gkiar** [comment]
Hey @prioux - to confirm, if I change it so that it is pulled (and therefore `locked`) in the directory where the image will eventually be stored, is that sufficient to resolve this issue? I have found the necessary lines to modify, I believe, and can make the patch

**Comment by prioux** [comment]
I'm not sure what you mean. Look, the command I run is very very simply:

```
bosh exec prepare --imagepath /path/to/something.img descriptor.json
```

and what I expect is really once all is done an over that have a file `/path/to/something.img` with my image in it. The lockdir should, of course, be `/path/to/something.img-lockdir` at all times because it has to be on the same filesystem as the image, and the tmp file while the image is being pull must also be on the same filesystem. If you can do that, then yes, go ahead.

**Comment by prioux** [comment]
@gkiar I think I understand your question and the answer is 'yes'. :-)

**Comment by gkiar** [comment]
@prioux -- what version of `boutiques` do you have installed, and for what version of Python? The latest version on the `develop` branch and Python3.6 is not giving me this issue, as shown below:

```bash
$ pwd  # Confirm my location
/home/users/gkiar/ace_mount/ace_home/code/gkiar/boutiques/tools/python
$ ls /home/users/gkiar/data/testdir  # list destination contents
$ bosh exec prepare --imagepath /home/users/gkiar/data/testdir/onevox.img ./oneVoxel.json  # pull image

$ ls /home/users/gkiar/data/testdir  # look in the destination directory while that is happening using another shell
onevox.img-lock  onevox.img.tmp

$ ls /home/users/gkiar/data/testdir  # look again once it's finished
onevox.img*
```

**Comment by prioux** [comment]
Have you tried with singularity 3.2 ? I think it might be the problem....

**Comment by prioux** [comment]
Yes, it seems so:

```
gra-platform1 ~ > which singularity
/opt/software/singularity-2.6/bin/singularity
gra-platform1 ~ > bosh version
0.5.19
gra-platform1 ~ > ls tst
gra-platform1 ~ > bosh exec prepare --imagepath $HOME/tst/abc.simg boutiques_demo.json 
Container location: Pulled from shub://BIDS-Apps/ndmg:v0.1.0 (abc.simg not found in current working directory or specified image path)
gra-platform1 ~ > module load singularity/3.2

The following have been reloaded with a version change:
  1) singularity/2.6 => singularity/3.2

gra-platform1 ~ > rm tst/abc.simg 
rm: remove regular file 'tst/abc.simg'? y
gra-platform1 ~ > bosh exec prepare --imagepath $HOME/tst/abc.simg boutiques_demo.json 
Traceback (most recent call last):
  File "/home/cbrain01/.local/bin/bosh", line 11, in <module>
    sys.exit(bosh())
  File "/home/cbrain01/.local/lib/python2.7/site-packages/boutiques/bosh.py", line 526, in bosh
    out = execute(*params)
  File "/home/cbrain01/.local/lib/python2.7/site-packages/boutiques/bosh.py", line 211, in execute
    container_location = executor.prepare()[1]
  File "/home/cbrain01/.local/lib/python2.7/site-packages/boutiques/localExec.py", line 450, in prepare
    conName, conIndex, conImage, imageDir, lockDir)
  File "/home/cbrain01/.local/lib/python2.7/site-packages/boutiques/localExec.py", line 496, in _pullSingImage
    os.rename(op.join(imageDir, conNameTmp), op.join(imageDir, conName))
OSError: [Errno 2] No such file or directory
gra-platform1 ~ > ls abc*
abc.simg.tmp
```

**Comment by prioux** [comment]
What I don't understand though is that when I run `singularity pull` with singularity 3.2, and I specify a path, it works perfectly fine.

I think bosh has some sort of code to handle different version of the pull command that singularity has supported in its history maybe?

**Comment by prioux** [comment]
Singularity changed. I'm using 3.2 on some servers and 2.6 on other.

With 2.6, the pull command is this (which is used in bosh)

```
singularity pull --name /path/to/name.simg shub://blah
```

With 3.2 it is

```
singularity pull /path/to/name.simg shub://blah
```

The 3.2 version doesn't document --name yet it supports it invisibly.

That being said, I still don't know why in bosh, when using 3.2, the pull happens in the current working directory. When I try running singularity pull directory with a path, it ends up in the proper place...

**Comment by natacha-beck** [comment]
@gkiar and @glatard the issue came from singularity https://github.com/sylabs/singularity/issues/2788. 

This functionality seems to be back in Singularity 3.3 and to be absent in 3.1 and 3.2 at least, these versions do not support the env variable "SINGULARITY_CACHEDIR"

Maybe we can change the following line: 
- https://github.com/boutiques/boutiques/blob/master/tools/python/boutiques/localExec.py#L473 

to include the `imageDir` when we pull the image it seems to be a workable solution at least for singularity 3.2.

**Comment by glatard** [comment]
At the moment we only support (i.e., test) singularity 2.6.1, because it used to be the only version available on Compute Canada. It looks like Compute Canada now has 3.4 .Shall we then move to 3.4 and ignore the (buggy) 3.1 and 3.2?

**Comment by glatard** [comment]
Will be solved in #560
